File: checker.cpp
/*
 * Password Strength Checker — C++ Implementation
 * DecodeLabs Industrial Training Kit | Batch 2026
 * Project 1 | Cybersecurity Track
 *
 * Compile:  g++ -std=c++17 -o checker checker.cpp
 * Run:      ./checker
 *
 * Mirrors the Python logic but in C++ — useful for seeing how
 * the same algorithm behaves across languages.
 * O(n) linear scan throughout — no regex, no heavy libraries.
 */

#include <iostream>
#include <string>
#include <vector>
#include <set>
#include <algorithm>
#include <cctype>
#include <iomanip>

// ─────────────────────────────────────────────
//  Common / leaked password list (small sample)
// ─────────────────────────────────────────────
const std::set<std::string> COMMON_PASSWORDS = {
    "password", "123456", "password123", "admin", "letmein",
    "qwerty", "abc123", "monkey", "1234567890", "superman",
    "iloveyou", "sunshine", "princess", "dragon", "master",
    "welcome", "login", "admin123", "pass", "test123",
    "password1", "12345678", "shadow", "football", "baseball"
};

// ANSI terminal colors
const std::string RED    = "\033[91m";
const std::string YELLOW = "\033[93m";
const std::string GREEN  = "\033[92m";
const std::string CYAN   = "\033[96m";
const std::string BOLD   = "\033[1m";
const std::string RESET  = "\033[0m";


// ─────────────────────────────────────────────
//  Result struct — mirrors Python dict output
// ─────────────────────────────────────────────
struct CheckResult {
    std::string strength;     // WEAK / MEDIUM / STRONG
    int         score;
    bool        has_upper;
    bool        has_lower;
    bool        has_digit;
    bool        has_symbol;
    bool        is_common;
    bool        has_repeats;
    int         length;
    std::vector<std::string> feedback;
};


// ─────────────────────────────────────────────
//  Helper: convert string to lowercase
// ─────────────────────────────────────────────
std::string to_lower(const std::string& s) {
    std::string result = s;
    std::transform(result.begin(), result.end(), result.begin(),
                   [](unsigned char c){ return std::tolower(c); });
    return result;
}


// ─────────────────────────────────────────────
//  Check: is password in common list?
// ─────────────────────────────────────────────
bool is_common_password(const std::string& password) {
    std::string lower = to_lower(password);
    return COMMON_PASSWORDS.count(lower) > 0;
}


// ─────────────────────────────────────────────
//  Check: three or more repeated consecutive chars
// ─────────────────────────────────────────────
bool has_repeated_chars(const std::string& password) {
    for (size_t i = 0; i + 2 < password.size(); ++i) {
        if (password[i] == password[i+1] && password[i+1] == password[i+2]) {
            return true;
        }
    }
    return false;
}


// ─────────────────────────────────────────────
//  Check: character variety (single linear pass)
// ─────────────────────────────────────────────
void check_variety(const std::string& password,
                   bool& has_upper, bool& has_lower,
                   bool& has_digit, bool& has_symbol) {
    has_upper  = false;
    has_lower  = false;
    has_digit  = false;
    has_symbol = false;

    // Symbols = printable chars that aren't letters or digits
    std::string symbols = "!@#$%^&*()-_=+[]{}|;:',.<>?/`~\"\\";

    for (char c : password) {
        if (std::isupper((unsigned char)c)) has_upper  = true;
        else if (std::islower((unsigned char)c)) has_lower  = true;
        else if (std::isdigit((unsigned char)c)) has_digit  = true;
        else if (symbols.find(c) != std::string::npos) has_symbol = true;
    }
}


// ─────────────────────────────────────────────
//  Master strength calculator
// ─────────────────────────────────────────────
CheckResult calculate_strength(const std::string& password) {
    CheckResult result;
    result.length = static_cast<int>(password.size());
    result.score  = 0;

    if (password.empty()) {
        result.strength = "INVALID";
        result.feedback.push_back("Password cannot be empty.");
        return result;
    }

    // --- Length check ---
    if (result.length < 8) {
        result.feedback.push_back("Too short — minimum 8 characters required.");
    } else if (result.length >= 16) {
        result.score += 3;
    } else if (result.length >= 12) {
        result.score += 2;
    } else {
        result.score += 1;
    }

    // --- Common password check ---
    result.is_common = is_common_password(password);
    if (result.is_common) {
        result.feedback.push_back("Commonly used password — easily cracked.");
        result.score = 0;
    }

    // --- Character variety ---
    check_variety(password,
                  result.has_upper, result.has_lower,
                  result.has_digit, result.has_symbol);

    int variety_score = (result.has_upper ? 1 : 0)
                      + (result.has_lower ? 1 : 0)
                      + (result.has_digit ? 1 : 0)
                      + (result.has_symbol ? 1 : 0);
    result.score += variety_score;

    if (!result.has_upper)  result.feedback.push_back("Add an uppercase letter (A-Z).");
    if (!result.has_digit)  result.feedback.push_back("Add a number (0-9).");
    if (!result.has_symbol) result.feedback.push_back("Add a symbol (e.g. @, #, !).");
    if (!result.has_lower)  result.feedback.push_back("Add a lowercase letter.");

    // --- Repeated chars ---
    result.has_repeats = has_repeated_chars(password);
    if (result.has_repeats) {
        result.score -= 1;
        result.feedback.push_back("Avoid repeating characters (e.g. 'aaa' or '111').");
    }

    result.score = std::max(0, result.score);

    // --- Final classification ---
    if (result.is_common || result.length < 8) {
        result.strength = "WEAK";
    } else if (result.score <= 3) {
        result.strength = "WEAK";
    } else if (result.score <= 5) {
        result.strength = "MEDIUM";
    } else {
        result.strength = "STRONG";
    }

    if (result.feedback.empty()) {
        result.feedback.push_back("Password looks solid. Good work!");
    }

    return result;
}


// ─────────────────────────────────────────────
//  Display result to terminal
// ─────────────────────────────────────────────
void print_result(const CheckResult& r, const std::string& password) {
    std::string color;
    std::string banner;

    if (r.strength == "STRONG") {
        color  = GREEN;
        banner = "✅  STRONG PASSWORD — Good to go";
    } else if (r.strength == "MEDIUM") {
        color  = YELLOW;
        banner = "⚠️   MEDIUM PASSWORD — Can be improved";
    } else {
        color  = RED;
        banner = "❌  WEAK PASSWORD — Not acceptable";
    }

    std::string masked(password.size(), '*');

    std::cout << "\n" << std::string(47, '-') << "\n";
    std::cout << "  Password   : " << masked          << "\n";
    std::cout << "  Length     : " << r.length         << " characters\n";
    std::cout << "  Score      : " << r.score          << " / 7\n";
    std::cout << "\n  " << BOLD << color << banner << RESET << "\n";
    std::cout << std::string(47, '-') << "\n";

    std::cout << "\n  Character Mix:\n";
    std::cout << "    Uppercase : " << (r.has_upper  ? "✓" : "✗") << "\n";
    std::cout << "    Lowercase : " << (r.has_lower  ? "✓" : "✗") << "\n";
    std::cout << "    Numbers   : " << (r.has_digit  ? "✓" : "✗") << "\n";
    std::cout << "    Symbols   : " << (r.has_symbol ? "✓" : "✗") << "\n";

    if (!r.feedback.empty()) {
        std::cout << "\n  Suggestions:\n";
        for (const auto& tip : r.feedback) {
            std::cout << "    → " << tip << "\n";
        }
    }

    std::cout << "\n" << std::string(47, '-') << "\n\n";
}


// ─────────────────────────────────────────────
//  Main — interactive CLI loop
// ─────────────────────────────────────────────
int main() {
    std::cout << "\n" << BOLD << CYAN;
    std::cout << "=======================================================\n";
    std::cout << "   🔐  PASSWORD STRENGTH CHECKER  (C++ Version)\n";
    std::cout << "   DecodeLabs | Cybersecurity Project 1\n";
    std::cout << "=======================================================\n";
    std::cout << RESET;
    std::cout << "  Type a password to check its strength.\n";
    std::cout << "  Type 'quit' to exit.\n\n";

    std::string password;

    while (true) {
        std::cout << "  Enter password: ";

        // Note: in a real system use platform-specific echo-off.
        // For this training project, normal getline is fine.
        if (!std::getline(std::cin, password)) break;

        if (password == "quit" || password == "exit" || password == "q") {
            std::cout << "\n  " << CYAN << "Exiting. Stay secure! 🔒" << RESET << "\n\n";
            break;
        }

        CheckResult result = calculate_strength(password);
        print_result(result, password);
    }

    return 0;
}

File: checker.py
"""
Password Strength Checker - Core Logic
DecodeLabs Industrial Training Kit | Batch 2026
Author: Intern | Project 1 - Cybersecurity Track

This module handles the actual strength evaluation logic.
Uses Python string methods and conditional checks — no regex overkill.
"""

import string
import hmac


# A small set of the most commonly used / leaked passwords
# In a real system you'd load from a file like rockyou.txt subset
COMMON_PASSWORDS = {
    "password", "123456", "password123", "admin", "letmein",
    "qwerty", "abc123", "monkey", "1234567890", "superman",
    "iloveyou", "sunshine", "princess", "dragon", "master",
    "welcome", "login", "admin123", "pass", "test123",
    "password1", "12345678", "shadow", "football", "baseball"
}


def check_length(password: str) -> dict:
    """
    Length is the first and most important check.
    < 8 chars = automatic fail (exponential brute-force risk)
    """
    length = len(password)
    if length < 8:
        return {"pass": False, "length": length, "note": "Too short — minimum 8 characters required"}
    elif length < 12:
        return {"pass": True, "length": length, "note": "Acceptable length"}
    elif length < 16:
        return {"pass": True, "length": length, "note": "Good length"}
    else:
        return {"pass": True, "length": length, "note": "Excellent length"}


def check_character_variety(password: str) -> dict:
    """
    Checks presence of:
    - Uppercase letters [A-Z]
    - Lowercase letters [a-z]
    - Digits [0-9]
    - Special symbols
    
    Uses Pythonic 'any()' approach with short-circuit evaluation
    instead of verbose manual loops.
    """
    has_upper  = any(c.isupper() for c in password)
    has_lower  = any(c.islower() for c in password)
    has_digit  = any(c.isdigit() for c in password)
    has_symbol = any(c in string.punctuation for c in password)

    score = sum([has_upper, has_lower, has_digit, has_symbol])

    return {
        "uppercase": has_upper,
        "lowercase": has_lower,
        "digit":     has_digit,
        "symbol":    has_symbol,
        "variety_score": score  # 0 to 4
    }


def check_common_password(password: str) -> bool:
    """
    Returns True if password is in the common/leaked list.
    Case-insensitive comparison.
    Uses hmac.compare_digest to avoid timing attacks
    when comparing against each entry.
    """
    pw_lower = password.lower()
    for common in COMMON_PASSWORDS:
        # constant-time comparison — prevents timing side-channel leaks
        if hmac.compare_digest(pw_lower, common):
            return True
    return False


def check_repeated_chars(password: str) -> bool:
    """
    Detects patterns like 'aaaa' or '1111' that reduce entropy.
    Simple sequential check — O(n) linear scan as required.
    """
    for i in range(len(password) - 2):
        if password[i] == password[i+1] == password[i+2]:
            return True  # Three or more same chars in a row
    return False


def calculate_strength(password: str) -> dict:
    """
    Master function — aggregates all checks and returns a result dict.
    
    Scoring logic:
      Weak   → fails length OR is a common password
      Medium → passes length, has 2-3 character varieties
      Strong → passes all checks, has 4 varieties, no repeated chars
    """
    if not password:
        return {
            "strength": "INVALID",
            "score": 0,
            "feedback": ["Password cannot be empty."],
            "details": {}
        }

    length_check  = check_length(password)
    variety_check = check_character_variety(password)
    is_common     = check_common_password(password)
    has_repeats   = check_repeated_chars(password)

    feedback = []
    score = 0

    # --- Length scoring ---
    if not length_check["pass"]:
        feedback.append(f"Too short ({length_check['length']} chars). Use at least 8.")
    else:
        if length_check["length"] >= 16:
            score += 3
        elif length_check["length"] >= 12:
            score += 2
        else:
            score += 1

    # --- Common password check ---
    if is_common:
        feedback.append("This is a commonly used password — easily cracked.")
        score = 0  # Hard reset; doesn't matter how complex it looks

    # --- Character variety scoring ---
    variety = variety_check["variety_score"]
    score += variety

    if not variety_check["uppercase"]:
        feedback.append("Add at least one uppercase letter (A-Z).")
    if not variety_check["digit"]:
        feedback.append("Add at least one number (0-9).")
    if not variety_check["symbol"]:
        feedback.append("Add at least one symbol (e.g. @, #, !).")
    if not variety_check["lowercase"]:
        feedback.append("Add at least one lowercase letter.")

    # --- Repeated character penalty ---
    if has_repeats:
        score -= 1
        feedback.append("Avoid repeating characters (e.g. 'aaa' or '111').")

    # --- Final classification ---
    if is_common or not length_check["pass"]:
        strength = "WEAK"
    elif score <= 3:
        strength = "WEAK"
    elif score <= 5:
        strength = "MEDIUM"
    else:
        strength = "STRONG"

    if not feedback:
        feedback.append("Password looks solid. Good work!")

    return {
        "strength": strength,
        "score": max(0, score),
        "feedback": feedback,
        "details": {
            "length":    length_check,
            "variety":   variety_check,
            "is_common": is_common,
            "has_repeats": has_repeats
        }
    }
